packages/flutter/lib/src/widgets/slotted_render_object_widget.dart

```diff
@@ -231,14 +241,43 @@ class SlottedRenderObjectElement<S> extends RenderObjectElement {
     }(), '${widget.runtimeType}.slots must not change.');
     assert(slottedMultiChildRenderObjectWidgetMixin.slots.toSet().length == slottedMultiChildRenderObjectWidgetMixin.slots.length, 'slots must be unique');
 
+    final Map<Key, S> oldKeyedSlots = _keyToSlot;
+    _keyToSlot = <Key, S>{};
+
+    // New slots for keyed children. It shouldn't be updated in place because
+    // two children could swap slots.
+    final Map<S, Element> slotToKeyedChild = <S, Element>{};
+    final List<MapEntry<S, Widget?>> unkeyedWidgets = <MapEntry<S, Widget?>>[];
+
     for (final S slot in slottedMultiChildRenderObjectWidgetMixin.slots) {
-      _updateChild(slottedMultiChildRenderObjectWidgetMixin.childForSlot(slot), slot);
+      final Widget? widget = slottedMultiChildRenderObjectWidgetMixin.childForSlot(slot);
+      final Key? newWidgetKey = widget?.key;
+      if (newWidgetKey != null) {
+        assert(!_keyToSlot.containsKey(newWidgetKey));
+        _keyToSlot[newWidgetKey] = slot;
+        final S fromSlot = oldKeyedSlots[newWidgetKey] ?? slot;
+        final Element newChild = updateChild(_slotToChild.remove(fromSlot), widget, slot)!;
+        assert(!_slotToChild.containsKey(fromSlot));
+        assert(!slotToKeyedChild.containsKey(slot));
+        slotToKeyedChild[slot] = newChild;
```

Comment:
Couldn't this (combined with some of the other comments) just re-use the logic in `_updateChild`?

Reply:
(We probably have to remove the assert in that method that wasn't valid anyways since you have to delete the justification comment there...)